src/shims/fs.rs

```diff
@@ -257,6 +264,186 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        // integer.
+        let flags: i32 = this.try_from_ptr_sized_operand(flags_op)?;
+
+        let at_fdcwd = this.eval_libc_i32("AT_FDCWD")?;
+        // Now we need to resolve the path of the target file. This was done following the manual
+        // entry for `statx`
+
+        // If `pathname` is absolute we can ignore `dirfd`.
+        let path = if pathname.is_absolute() {
+            pathname
+        } else {
+            // This argument should be a `c_int` but the syscall API just returns a pointer sized
+            // integer.
+            let dirfd: i32 = this.try_from_ptr_sized_operand(dirfd_op)?;
+            // If `dirfd` is `AT_FDCWD`, `pathname` is relative to the current directory.
+            if dirfd == at_fdcwd {
+                match std::env::current_dir() {
+                    Ok(cwd) => cwd.join(pathname),
+                    Err(e) => {
+                        this.set_last_error_from_io_error(e)?;
+                        return Ok(-1);
+                    }
+                }
+            } else {
+                // This behavior is specified but cannot be tested from `libstd`. If you found this
+                // error please open an issue reporting it.
+                throw_unsup_format!(
+                    "Using statx with a fildes different from `AT_FDCWD` is not supported"
+                )
+            }
+        };
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
+            // FIXME: metadata for symlinks need testing.
+            std::fs::symlink_metadata(path)
+        } else {
+            std::fs::metadata(path)
+        };
+
+        let metadata = match metadata {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        // The `mode` field specifies the type of the file and the permissions over the file for
+        // the owner, its group and other users. Given that we can only provide the file type
+        // without using platform specific methods, we only set the bits corresponding to the file
+        // type. This argument should be an `__u16` but the syscall API just returns a pointer sized
```

Comment:
"the syscall API" doesn't *return* anything here, does it?

Reply:
No, that's just bad wording from my part. "provides" maybe?

Reply:
Thus is not about "the syscall API" at all. The thing is that `syscall` is an *untyped* function so all types are determined by the caller, and the particular caller we are working with here decided to make everything an `isize`. (I am still trying to figure out why.)
This is not a universal truth for all syscalls.